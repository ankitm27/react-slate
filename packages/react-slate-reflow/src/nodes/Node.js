/* @flow */

import assert from 'assert';
import type {
  Child,
  Parent,
  LayoutProps,
  StyleProps,
  BorderProps,
  Traversable,
} from '../types';

export default class Node implements Traversable<Child> {
  parent: ?Parent = null;
  children: Array<Traversable<Child>> = [];

  layoutProps: ?LayoutProps = null;
  styleProps: ?StyleProps = null;
  borderProps: ?BorderProps = null;

  constructor({ parent }: { parent: Node } = {}) {
    if (parent) {
      this.parent = parent;
    }
  }

  setLayoutProps(layoutProps: ?LayoutProps) {
    this.layoutProps = layoutProps;
  }

  setStyleProps(styleProps: ?StyleProps) {
    this.styleProps = styleProps;
  }

  setBorder(borderProps: ?BorderProps) {
    this.borderProps = borderProps;
    // $FlowFixMe
    assert.fail('missing implementation');
    // TODO: transform children tree to account for border
  }

  prependChild(child: Child, childBefore: Child) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    const index = this.children.indexOf(childBefore);
    this.children.splice(index, 0, child);
  }

  appendChild(child: Child) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    this.children.push(child);
  }

  insertChild(child: Child, position?: number) {
    const index =
      typeof position !== 'undefined' ? position : this.children.length;

    assert(
      index <= this.children.length + 1,
      `child position out of bounds: ${index}`
    );

    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    this.children[index] = child;
  }

  removeChild(child: Child) {
    const index = this.children.indexOf(child);

    assert(index >= 0, 'child not found');

    // eslint-disable-next-line no-param-reassign
    child.parent = null;
    this.children.splice(index, 1);
  }
}
