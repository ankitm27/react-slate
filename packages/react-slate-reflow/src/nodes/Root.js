/* @flow */

import assert from 'assert';
import type { Child, Traversable } from '../types';
import calculateLayout from '../layout/calculateLayout';

type Size = {
  width: number,
  height: number,
};

export default class Root implements Traversable<Child> {
  children: Array<Traversable<Child>> = [];
  size: Size;

  constructor(size: Size) {
    this.size = size;
  }

  prependChild(child: Child, childBefore: Child) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    const index = this.children.indexOf(childBefore);
    this.children.splice(index, 0, child);
  }

  appendChild(child: Child) {
    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    this.children.push(child);
  }

  insertChild(child: Child, position?: number) {
    const index =
      typeof position !== 'undefined' ? position : this.children.length;

    assert(index <= this.children.length + 1, 'child position out of bounds');

    // eslint-disable-next-line no-param-reassign
    child.parent = this;
    this.children[index] = child;
  }

  removeChild(child: Child) {
    const index = this.children.indexOf(child);

    assert(
      index >= 0 && index < this.children.length,
      `child position out of bounds: ${index}`
    );

    // eslint-disable-next-line no-param-reassign
    child.parent = null;

    this.children.splice(index, 1);
  }

  calculateLayout() {
    return calculateLayout(this);
  }
}
