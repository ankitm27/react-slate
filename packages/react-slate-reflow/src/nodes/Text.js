/* @flow */

import type { Parent } from '../types';
import type { Traversable } from '../lib/BinaryTree';

export default class Text implements Traversable<*> {
  body: string = '';
  parent: ?Parent = null;
  // Children should always be empty.
  children: Array<*> = Object.freeze([]);

  setBody(body: string) {
    this.body = body;
  }
}
