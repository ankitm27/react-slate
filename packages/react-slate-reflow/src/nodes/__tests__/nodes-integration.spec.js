import Root from '../Root';
import Node from '../Node';
import Text from '../Text';

describe('nodes integration suite', () => {
  describe('should calculate layout', () => {
    describe('for node -> text', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const node = new Node();
        const text = new Text();

        text.setBody('Hello World');
        node.insertChild(text);
        root.insertChild(node);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello World',
        //       style: null,
        //     },
        //   },
        // ]);
      });

      xit('with style and layout props', () => {
        const root = getTree();

        root.children[0].setLayoutProps({
          marginLeft: 2,
          marginTop: 1,
          paddingTop: 1,
        });
        root.children[0].setStyleProps({
          backgroundColor: 'red',
          color: 'blue',
        });

        const { elements } = root.calculateLayout();
        expect(elements).toEqual([
          {
            box: {
              style: {
                backgroundColor: 'red',
                color: 'blue',
              },
              x: 2,
              y: 1,
              width: 20,
              height: 2,
            },
          },
          {
            body: {
              x: 2,
              y: 2,
              value: 'Hello World',
              style: {
                color: 'blue',
              },
            },
          },
        ]);
      });
    });

    describe('for node -> node -> text', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const innerNode = new Node();
        const text = new Text();

        text.setBody('Hello World');
        innerNode.insertChild(text);
        outerNode.insertChild(innerNode);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello World',
        //       style: null,
        //     },
        //   },
        // ]);
      });

      xit('with style and layout props', () => {
        const root = getTree();

        root.children[0].setLayoutProps({
          marginTop: 1,
          paddingTop: 1,
        });
        root.children[0].setStyleProps({ backgroundColor: 'red' });

        root.children[0].children[0].setLayoutProps({
          marginLeft: 2,
          marginRight: 2,
        });
        root.children[0].children[0].setStyleProps({ backgroundColor: 'blue' });

        const { elements } = root.calculateLayout();

        expect(elements).toEqual([
          {
            box: {
              style: {
                backgroundColor: 'red',
              },
              x: 0,
              y: 1,
              width: 20,
              height: 2,
            },
          },
          {
            box: {
              style: {
                backgroundColor: 'blue',
              },
              x: 2,
              y: 2,
              width: 20,
              height: 1,
            },
          },
          {
            body: {
              x: 2,
              y: 2,
              value: 'Hello World',
              style: null,
            },
          },
        ]);
      });
    });

    describe('for node -> [text, text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const node = new Node();
        const text1 = new Text();
        const text2 = new Text();

        text1.setBody('Hello');
        text2.setBody(' World');
        node.insertChild(text1);
        node.insertChild(text2);
        root.insertChild(node);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 5,
        //       y: 0,
        //       value: ' World',
        //       style: null,
        //     },
        //   },
        // ]);
      });

      xit('with style and layout props', () => {
        const root = getTree();

        root.children[0].setLayoutProps({
          marginLeft: 2,
          marginTop: 2,
        });
        root.children[0].setStyleProps({ backgroundColor: 'blue' });

        const { elements } = root.calculateLayout();

        expect(elements).toEqual([
          {
            box: {
              style: {
                backgroundColor: 'blue',
              },
              x: 2,
              y: 2,
              width: 20,
              height: 1,
            },
          },
          {
            body: {
              x: 2,
              y: 2,
              value: 'Hello',
              style: null,
            },
          },
          {
            body: {
              x: 7,
              y: 2,
              value: ' World',
              style: null,
            },
          },
        ]);
      });
    });

    describe('for node -> [text, node -> text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const text = new Text();
        const innerNode = new Node();
        const innerText = new Text();

        text.setBody('Hello');
        innerText.setBody('World');
        innerNode.insertChild(innerText);
        outerNode.insertChild(text);
        outerNode.insertChild(innerNode);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 1,
        //       value: 'World',
        //       style: null,
        //     },
        //   },
        // ]);
      });
    });

    describe('for node -> [node -> text, text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const text = new Text();
        const innerNode = new Node();
        const innerText = new Text();

        text.setBody('World');
        innerText.setBody('Hello');
        innerNode.insertChild(innerText);
        outerNode.insertChild(innerNode);
        outerNode.insertChild(text);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 1,
        //       value: 'World',
        //       style: null,
        //     },
        //   },
        // ]);
      });
    });

    describe('for node -> [node -> text, node -> text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const innerNode1 = new Node();
        const text1 = new Text();
        const innerNode2 = new Node();
        const text2 = new Text();

        text1.setBody('Hello');
        text2.setBody('World');
        innerNode1.insertChild(text1);
        innerNode2.insertChild(text2);
        outerNode.insertChild(innerNode1);
        outerNode.insertChild(innerNode2);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Hello',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 1,
        //       value: 'World',
        //       style: null,
        //     },
        //   },
        // ]);
      });
    });

    describe('for node -> [text, node -> text, text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const innerNode = new Node();
        const text1 = new Text();
        const text2 = new Text();
        const text3 = new Text();

        text1.setBody('Brave');
        text2.setBody('New');
        text3.setBody('World');
        innerNode.insertChild(text2);
        outerNode.insertChild(text1);
        outerNode.insertChild(innerNode);
        outerNode.insertChild(text3);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Brave',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 1,
        //       value: 'New',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 2,
        //       value: 'World',
        //       style: null,
        //     },
        //   },
        // ]);
      });
    });

    describe('for node -> [node -> text, text, node -> text]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();
        const innerNode1 = new Node();
        const innerNode2 = new Node();
        const text1 = new Text();
        const text2 = new Text();
        const text3 = new Text();

        text1.setBody('Brave');
        text2.setBody('New');
        text3.setBody('World');
        innerNode1.insertChild(text1);
        innerNode2.insertChild(text3);
        outerNode.insertChild(innerNode1);
        outerNode.insertChild(text2);
        outerNode.insertChild(innerNode2);
        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
        // expect(elements).toEqual([
        //   {
        //     body: {
        //       x: 0,
        //       y: 0,
        //       value: 'Brave',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 1,
        //       value: 'New',
        //       style: null,
        //     },
        //   },
        //   {
        //     body: {
        //       x: 0,
        //       y: 2,
        //       value: 'World',
        //       style: null,
        //     },
        //   },
        // ]);
      });
    });

    describe('for node -> [node -> text, text, node -> [text, text]]', () => {
      function getTree() {
        const root = new Root({ width: 20, height: 10 });
        const outerNode = new Node();

        const node1 = new Node();
        const text1 = new Text();
        node1.insertChild(text1);
        text1.setBody('Text1');

        const text2 = new Text();
        text2.setBody('Text2');

        const node2 = new Node();
        const text3 = new Text();
        const text4 = new Text();
        text3.setBody('Text3');
        text4.setBody('Text4');
        node2.insertChild(text3);
        node2.insertChild(text4);

        outerNode.insertChild(node1);
        outerNode.insertChild(text2);
        outerNode.insertChild(node2);

        root.insertChild(outerNode);

        return root;
      }

      it('without style/layout props', () => {
        const root = getTree();
        const { layoutTree } = root.calculateLayout();
        expect(layoutTree.getJsonTree()).toMatchSnapshot();
      });
    });
  });

  xit('node -> [node (bg) -> text, node -> text]', () => {
    const root = new Root({ width: 20, height: 10 });
    const parentNode = new Node();

    const node1 = new Node();
    const text1 = new Text();
    text1.setBody('Hello');
    node1.insertChild(text1);
    node1.setLayoutProps({
      marginLeft: 2,
      paddingBottom: 1,
      paddingTop: 1,
    });
    node1.setStyleProps({ backgroundColor: 'red' });

    const node2 = new Node();
    const text2 = new Text();
    text2.setBody('World');
    node2.insertChild(text2);
    node2.setLayoutProps({
      marginTop: 2,
    });

    parentNode.insertChild(node1);
    parentNode.insertChild(node2);
    root.insertChild(parentNode);

      debugger // eslint-disable-line
    const { elements } = root.calculateLayout();

    expect(elements).toEqual([
      {
        box: {
          style: {
            backgroundColor: 'red',
          },
          x: 2,
          y: 0,
          width: 20,
          height: 3,
        },
      },
      {
        body: {
          x: 2,
          y: 1,
          value: 'Hello',
          style: null,
        },
      },
      {
        body: {
          x: 0,
          y: 5,
          value: 'World',
          style: null,
        },
      },
    ]);
  });

  // xit('should calculate layout', () => {
  //   const root = new Root({ width: 20, height: 10 });

  //   const node1 = new Node();
  //   const node11 = new Text();
  //   node11.setBody('Hello World');
  //   node1.insertChild(node11);
  //   node1.setLayoutProps({
  //     marginLeft: 2,
  //     marginRight: 2,
  //     marginBottom: 1,
  //   });

  //   const node2 = new Node();
  //   const node21 = new Node();
  //   const node211 = new Text();
  //   node211.setBody('react-slate');
  //   node21.insertChild(node211);
  //   node21.setStyleProps({ backgroundColor: 'ansi-red' });
  //   node2.insertChild(node21);
  //   node2.setLayoutProps({
  //     paddingTop: 1,
  //     paddingBottom: 1,
  //   });

  //   const node3 = new Text();
  //   node3.setBody('Lorem ipsum');

  //   root.insertChild(node1);
  //   root.insertChild(node2);
  //   root.insertChild(node3);

  //   debugger // eslint-disable-line

  //   const elements = root.calculateLayout();
  //   console.log(JSON.stringify(elements, null, '  '));
  //   //   expect(elements).toEqual([
  //   //     {
  //   //       body: {
  //   //         value: 'Hello World',
  //   //         x: 0,
  //   //         y: 0,
  //   //       },
  //   //       styleProps: null,
  //   //       x: 2,
  //   //       y: 0,
  //   //       width: 11,
  //   //       height: 1,
  //   //     },
  //   //     {
  //   //       body: {
  //   //         value: 'react-slate',
  //   //         x: 0,
  //   //         y: 1,
  //   //       },
  //   //       styleProps: { backgroundColor: 'ansi-red' },
  //   //       x: 0,
  //   //       y: 2,
  //   //       width: 11,
  //   //       height: 3,
  //   //     },
  //   //     {
  //   //       body: {
  //   //         value: 'Lorem ipsum',
  //   //         x: 0,
  //   //         y: 0,
  //   //       },
  //   //       styleProps: null,
  //   //       x: 12,
  //   //       y: 4,
  //   //       width: 11,
  //   //       height: 1,
  //   //     },
  //   //   ]);
  // });

  // xit('[node (bg) -> text, node -> text]', () => {
  //   const root = new Root({ width: 20, height: 10 });

  //   const node1 = new Node();
  //   const text1 = new Text();
  //   text1.setBody('Hello');
  //   node1.insertChild(text1);
  //   node1.setLayoutProps({
  //     marginLeft: 2,
  //     paddingBottom: 1,
  //     paddingTop: 1,
  //   });
  //   node1.setStyleProps({ backgroundColor: 'red' });

  //   const node2 = new Node();
  //   const text2 = new Text();
  //   text2.setBody('World');
  //   node2.insertChild(text2);
  //   node2.setLayoutProps({
  //     marginTop: 2,
  //   });

  //   root.insertChild(node1);
  //   root.insertChild(node2);

  //   const { elements } = root.calculateLayout();

  //   expect(elements).toEqual([
  //     {
  //       box: {
  //         style: {
  //           backgroundColor: 'red',
  //         },
  //         x: 2,
  //         y: 0,
  //         width: 20,
  //         height: 3,
  //       },
  //     },
  //     {
  //       body: {
  //         x: 2,
  //         y: 1,
  //         value: 'Hello',
  //         style: null,
  //       },
  //     },
  //     {
  //       body: {
  //         x: 0,
  //         y: 5,
  //         value: 'World',
  //         style: null,
  //       },
  //     },
  //   ]);
  // });
});
