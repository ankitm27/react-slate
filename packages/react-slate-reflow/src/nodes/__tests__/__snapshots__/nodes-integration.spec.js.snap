// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodes integration suite should calculate layout for node -> [node -> text, node -> text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "body": "Hello",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 0,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "BlockLayout",
        },
        Object {
          "children": Array [
            Object {
              "body": "World",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 1,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "BlockLayout",
        },
      ],
      "dimensions": Object {
        "height": 2,
        "width": 5,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 2,
    "width": 5,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [node -> text, text, node -> [text, text]] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "body": "Text1",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 0,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "BlockLayout",
        },
        Object {
          "body": "Text2",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "InlineLayout",
        },
        Object {
          "children": Array [
            Object {
              "body": "Text3",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 2,
              },
              "type": "InlineLayout",
            },
            Object {
              "body": "Text4",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 5,
                "y": 2,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 10,
          },
          "placement": Object {
            "x": 0,
            "y": 2,
          },
          "type": "BlockLayout",
        },
      ],
      "dimensions": Object {
        "height": 3,
        "width": 10,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 3,
    "width": 10,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [node -> text, text, node -> text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "body": "Brave",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 0,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "BlockLayout",
        },
        Object {
          "body": "New",
          "dimensions": Object {
            "height": 1,
            "width": 3,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "InlineLayout",
        },
        Object {
          "children": Array [
            Object {
              "body": "World",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 2,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 2,
          },
          "type": "BlockLayout",
        },
      ],
      "dimensions": Object {
        "height": 3,
        "width": 5,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 3,
    "width": 5,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [node -> text, text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "body": "Hello",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 0,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "BlockLayout",
        },
        Object {
          "body": "World",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "InlineLayout",
        },
      ],
      "dimensions": Object {
        "height": 2,
        "width": 5,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 2,
    "width": 5,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [text, node -> text, text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "Brave",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "InlineLayout",
        },
        Object {
          "children": Array [
            Object {
              "body": "New",
              "dimensions": Object {
                "height": 1,
                "width": 3,
              },
              "placement": Object {
                "x": 0,
                "y": 1,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 3,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "BlockLayout",
        },
        Object {
          "body": "World",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 2,
          },
          "type": "InlineLayout",
        },
      ],
      "dimensions": Object {
        "height": 3,
        "width": 5,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 3,
    "width": 5,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [text, node -> text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "Hello",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "InlineLayout",
        },
        Object {
          "children": Array [
            Object {
              "body": "World",
              "dimensions": Object {
                "height": 1,
                "width": 5,
              },
              "placement": Object {
                "x": 0,
                "y": 1,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 1,
          },
          "type": "BlockLayout",
        },
      ],
      "dimensions": Object {
        "height": 2,
        "width": 5,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 2,
    "width": 5,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> [text, text] without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "Hello",
          "dimensions": Object {
            "height": 1,
            "width": 5,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "InlineLayout",
        },
        Object {
          "body": " World",
          "dimensions": Object {
            "height": 1,
            "width": 6,
          },
          "placement": Object {
            "x": 5,
            "y": 0,
          },
          "type": "InlineLayout",
        },
      ],
      "dimensions": Object {
        "height": 1,
        "width": 11,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 1,
    "width": 11,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> node -> text without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "body": "Hello World",
              "dimensions": Object {
                "height": 1,
                "width": 11,
              },
              "placement": Object {
                "x": 0,
                "y": 0,
              },
              "type": "InlineLayout",
            },
          ],
          "dimensions": Object {
            "height": 1,
            "width": 11,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "BlockLayout",
        },
      ],
      "dimensions": Object {
        "height": 1,
        "width": 11,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 1,
    "width": 11,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;

exports[`nodes integration suite should calculate layout for node -> text without style/layout props 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "body": "Hello World",
          "dimensions": Object {
            "height": 1,
            "width": 11,
          },
          "placement": Object {
            "x": 0,
            "y": 0,
          },
          "type": "InlineLayout",
        },
      ],
      "dimensions": Object {
        "height": 1,
        "width": 11,
      },
      "placement": Object {
        "x": 0,
        "y": 0,
      },
      "type": "BlockLayout",
    },
  ],
  "dimensions": Object {
    "height": 1,
    "width": 11,
  },
  "placement": Object {
    "x": 0,
    "y": 0,
  },
  "type": "BlockLayout",
}
`;
